src/irmin-pack/unix/gc_worker.ml

```diff
@@ -267,57 +278,82 @@ module Make (Args : Gc_args.S) = struct
-        |> Errs.log_if_error "GC: Close suffix")
-    @@ fun () ->
-    let append_exn = Ao.append_exn suffix in
-
-    (* Step 7. Transfer to the next suffix. *)
-    [%log.debug "GC: transfering to the new suffix"];
-    stats := Gc_stats.Worker.finish_current_step !stats "suffix: transfer";
-    let num_iterations = 7 in
-    (* [transfer_loop] is needed because after garbage collection there may be new objects
-       at the end of the suffix file that need to be copied over *)
-    let rec transfer_loop i ~off =
-      if i = 0 then off
-      else
-        let () = Fm.reload fm |> Errs.raise_if_error in
-        let pl : Payload.t = Fm.Control.payload (Fm.control fm) in
-        let end_offset =
-          Dispatcher.offset_of_suffix_poff dispatcher pl.suffix_end_poff
-        in
-        let len = Int63.Syntax.(end_offset - off) in
-        [%log.debug
-          "GC: transfer_loop iteration %d, offset %a, length %a"
-            (num_iterations - i + 1)
-            Int63.pp off Int63.pp len];
-        stats := Gc_stats.Worker.add_suffix_transfer !stats len;
-        let () = Dispatcher.read_bytes_exn dispatcher ~f:append_exn ~off ~len in
-        (* Check how many bytes are left, [4096*5] is selected because it is roughly the
-           number of bytes that requires a read from the block device on ext4 *)
-        if Int63.to_int len < 4096 * 5 then end_offset
-        else
-          let off = Int63.Syntax.(off + len) in
-          transfer_loop ~off (i - 1)
-    in
-    let new_end_suffix_offset =
-      transfer_loop ~off:new_prefix_end_offset num_iterations
+    (* Step 6. Calculate new suffix values. *)
+    let new_suffix, chunk_idx_gced =
+      stats :=
+        Gc_stats.Worker.finish_current_step !stats
+          "suffix: calculate new values";
+      let suffix = Fm.suffix fm in
+      let soff =
+        Dispatcher.suffix_poff_of_offset dispatcher new_suffix_start_offset
+      in
+      (* Step 6.1. Calculate chunks that we have GCed. *)
+      let open struct
+        type chunk = { idx : int; end_suffix_off : int63 }
+      end in
+      let chunks_gced =
+        match Fm.Suffix.chunk_num suffix with
+        | 1 -> [] (* We never remove a single chunk. *)
+        | _ ->
+            Fm.Suffix.fold_chunks
+              (fun ~acc ~idx ~start_suffix_off ~end_suffix_off ~is_appendable ->
+                (* Remove chunks that end at or before our new split point.
+                   This will leave the chunk that starts with (or contains) the
+                   split point but remove all previous chunks.
+                   We never remove empty or appendable chunks. *)
+                let is_empty = start_suffix_off = end_suffix_off in
+                let ends_with_or_is_before_soff = end_suffix_off <= soff in
+                let is_removable =
+                  (not is_appendable)
+                  && (not is_empty)
+                  && ends_with_or_is_before_soff
+                in
```

Comment:
IINM, an empty chunk is always appendable. You could remove `is_empty` from the condition and maybe salvage it into an `assert (if is_empty then is_appendable)`

Reply:
Hmm, that is true. We could remove `is_empty` since other parts of the code protect (added in the split PR) against an empty chunk being not appendable, but I'd prefer to not rely on that from the perspective of the GC code. I like that the current construction of the code independently answers the question, What chunks are removable?

`A chunk is removable if it is not appendable, not empty, and ends with or is before soff`

An assert would cover programmer error that allows in an empty, non-appendable chunk, but I think an assert would make it less readable. WDYT?

Reply:
> I'd prefer to not rely on that from the perspective of the GC code

Sure. Let's leave it like that